src/NHibernate/Util/AsyncReaderWriterLock.cs

```diff
@@ -182,19 +189,15 @@ private void ExitReadLock()
 			}
 		}
 
-		private void ReleaseWaitingReader()
+		private void DisposeWaitingSemaphore()
 		{
-			lock (_writeLockSemaphore)
+			if (_waitingDisposalSemaphore == null)
 			{
-				_readersWaiting--;
-				if (_readersWaiting != 0)
-				{
-					return;
-				}
-
-				_waitingReadLockSemaphore.Dispose();
-				_waitingReadLockSemaphore = null;
+				return;
 			}
+
+			_waitingDisposalSemaphore.Dispose();
```

Comment:
Using `_waitingDisposalSemaphore?.Dispose();` would be safer in principle (since we are in a threaded class), and simplifies the code by removing the `if`.

Reply:
True, fixed.